wrappers/core18.go

```diff
@@ -657,6 +661,50 @@ func undoSnapdDbusActivationOnCore() error {
 	return err
 }
 
+func snapdDesktopFileNames(s *snap.Info) ([]string, error) {
+	paths, err := filepath.Glob(filepath.Join(s.MountDir(), "usr/share/applications/*.desktop"))
+	if err != nil {
+		return nil, err
+	}
+
+	fileNames := make([]string, len(paths))
+	for i, path := range paths {
+		fileNames[i] = filepath.Base(path)
+	}
+
+	return fileNames, nil
+}
+
+func writeSnapdDesktopFilesOnCore(s *snap.Info) error {
+	// Ensure /var/lib/snapd/desktop/applications exists
+	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
+		return err
+	}
+
+	desktopFileNames, err := snapdDesktopFileNames(s)
+	if err != nil {
+		return err
+	}
+
+	desktopFiles := make(map[string]osutil.FileState, len(desktopFileNames))
+	for _, fileName := range desktopFileNames {
+		desktopFiles[fileName] = &osutil.FileReference{Path: filepath.Join(s.MountDir(), "usr/share/applications", fileName)}
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDesktopFilesDir, desktopFileNames, desktopFiles)
+	return err
+}
+
+func undoSnapdDesktopFilesOnCore(s *snap.Info) error {
```

Comment:
I am not sure if we need to add a test to cover the case where the snapd desktop files are changed. For example:
```
io.snapcraft.SessionAgent.desktop -> unchanged
snap-handle-link.desktop -> changed to snap-handle-link-v2.desktop
```
In this case `snap-handle-link.desktop` should be removed, but it won't.

Reply:
I don't understand - why would the desktop files have changed?

Reply:
You are right, normally they shouldn't, I was just thinking of the (admittedly rare) scenario where any of these desktop files would be deprecated for some reason in the future.

Maybe we should postpone this until/if it is needed? What do you think?

Reply:
Perhaps the right approach would be to hard code the list of desktop files similar to `dbusSessionServices` rather than listing files in the current snapd snap. This way we could keep a filename in the list after a rename/deletion so the `EnsureDirStateGlobs` call can remove it.

Reply:
Updated to match the dbus code.